Documentation/jdfs-config.1.in
.TH JDFS-CONFIG 1 2024-09-13 "jdfs @tag@"
.
.SH NAME
.
jdfs\-config \- get and set configuration options of the jdfs utility
.
.SH SYNOPSIS
.
.SY "jdfs config"
.RB { " \-h " | " \-\-help " }
.YS
.
.SY "jdfs config"
.RB { " \-g " | " \-\-get \fIoption " }
.SY "jdfs config"
.RB { " \-l " | " \-\-list " }
.SY "jdfs config"
.RB { " \-s " | " \-\-set \fIoption " } " \fIvalue"
.SY "jdfs config"
.RB { " \-u " | " \-\-unset \fIoption " }
.YS
.
.SH DESCRIPTION
.
.P
The \fBjdfs-config\fP command is the preferred way to get, set and unset
configuration options of the \fBjdfs\fP utility. It is, of course, also possible
to set configuration options by editing the configuration file directly (see
\fIFILES\fP), but using \fBjdfs-config\fP is way less prone to errors, so it is
strongly recommended that you use it.
.P
If no arguments are specified, \fBjdfs-config\fP will simply list all
configuration options and their values, behaving essentially as if \fB--list\fP
were specified.
.P
It is also worth mentioning that \fBjdfs-config\fP will allow you to set any
configuration option, even if the \fBjdfs\fP utility doesn't do anything with it.
To see which configuration options are actually read and used by \fBjdfs\fP,
please refer to
.BR jdfs.conf (5).
.P
Note: Configuration option names can only contain up to 63 bytes. Configuration
values, on the other hand, can contain up to 255 bytes. Beware these limits,
especially if there are non-ASCII characters in either name or value, as
non-ASCII characters may take up more than one byte.
.
.SH OPTIONS
.
.TP
.B \-h, \-\-help
Display a short help message.
.
.TP
.BI "\-g, \-\-get " option
Display the value of a configuration option.
.
.TP
.B \-l, \-\-list
Display the values of all options set in the config file.
.
.TP
.BI "\-s, \-\-set " "option value"
Set a configuration option to a value. Please note that this will create the
configuration file (see \fIFILES\fP) and all its parent directories, if it
doesn't exist yet.
.
.TP
.BI "\-u, \-\-unset " option
Unset a configuration option.
.
.P
Please note that only the first of the above arguments will be read. The
following command will, for example, only unset the \fItest\fP option, and not
display the value of \fIjdfs_root\fP.
.P
.IP
.EX
$ jdfs config -u test -g jdfs_root
.EE
.
.SH EXIT STATUS
.
.TP
.B 0
SUCCESS: Everything went as it should.
.
.TP
.B 1
ERROR: An error outside of below categories occurred.
.
.TP
.B 2
NOT_FOUND: The target of a search operation was not found.
.
.TP
.B 3
INPUT_ERROR: Invalid input was provided by the user.
.
.TP
.B 4
CONFIG_ERROR: An error occurred in the configuration mechanism.
.
.SH ENVIRONMENT
.
.TP
.I XDG_CONFIG_HOME
Path to the user-specific configuration directory. Defaults to
\fI\(ti/.config\fP. See \fIFILES\fP.
.
.SH FILES
.
.TP
.I $XDG_CONFIG_HOME/.config/jdfs/jdfs.conf
This is the main \fBjdfs\fP configuration file. See
.BR jdfs.conf (5).
.
.SH EXAMPLES
.
.P
Show the help message:
.IP
.EX
jdfs config --help
.EE
.
.P
Set the configuration option "jdfs_root" to \fI\(ti/jd\fP:
.IP
.EX
jdfs config --set jdfs_root \(ti/jd
.EE
.
.P
List all configured options:
.IP
.EX
jdfs config --list
.EE
.
.P
Display the value of the configuration option "show_hidden":
.IP
.EX
jdfs config --get show_hidden
.EE
.
.P
Unset the configuration option "show_hidden":
.IP
.EX
jdfs config --unset show_hidden
.EE
.
.SH SEE ALSO
.
.nh
.BR jdfs (1),
.BR jdfs-ls (1),
.BR jdfs-path (1),
.BR jdfs.conf (5)
.
.P
.UR https://johnnydecimal.com
.UE


src/jdfs-config.c
// Copyright (C) 2024 Timo Früh
// See main.c for the full notice.

#include "jdfs-config.h"

enum actions {
    CONFIG_HELP,
    CONFIG_GET,
    CONFIG_LIST,
    CONFIG_SET,
    CONFIG_UNSET,
    CONFIG_NULL
} action = CONFIG_LIST;

struct option longopts[] = {
    { "help", no_argument, NULL, 'h' },
    { "get", required_argument, NULL, 'g' },
    { "list", no_argument, NULL, 'l' },
    { "set", required_argument, NULL, 's' },
    { "unset", required_argument, NULL, 'u' },
    { NULL, 0, NULL, 0 }
};

void config_print_help(char argv_0[]) {

    char help_string[] =
        "Usage:\n"
        "\n"
        "  jdfs %1$s [ -h | --help ]\n"
        "  jdfs %1$s [ -g | --get ] option\n"
        "  jdfs %1$s [ -l | --list ]\n"
        "  jdfs %1$s [ -s | --set ] option value\n"
        "  jdfs %1$s [ -u | --unset ] option\n"
        "\n"
        "For more information, please consult the manual page.\n";

    printf(help_string, argv_0);
}

int config_set(char* name, char* value, char* path) {
    if (value == NULL) {
        snprintf(error_str, ERROR_STR_BUFSIZE, "%s couldn't be set: empty value", name);
        return INPUT_ERROR;
    }

    char swappath[CONFIG_PATH_BUFSIZE+1];
    snprintf(swappath, CONFIG_PATH_BUFSIZE+1, "%s~", path);

    return write_conf_data(name, value, path, swappath);
};

int config_get(char* name, char* path) {

    char value_buffer[CONFIG_VALUE_BUFSIZE];

    int read_retval = read_conf_data(value_buffer, CONFIG_VALUE_BUFSIZE, path, name);

    if (read_retval != SUCCESS) {
        return read_retval;
    }

    printf("%s\n", value_buffer);

    return SUCCESS;

}

int config_unset(char* name, char* path) {
    char swappath[CONFIG_PATH_BUFSIZE+1];
    snprintf(swappath, CONFIG_PATH_BUFSIZE+1, "%s~", path);

    return delete_conf_data(name, path, swappath);
}

int config_optswitch(int optchar, char* optarg) {
    switch (optchar) {
        case 'h':
            action = CONFIG_HELP;
            return OPT_FINAL;

        case 'g':
            action = CONFIG_GET;
            return OPT_FINAL;

        case 'l':
            action = CONFIG_LIST;
            return OPT_FINAL;

        case 's':
            action = CONFIG_SET;
            return OPT_FINAL;

        case 'u':
            action = CONFIG_UNSET;
            return OPT_FINAL;

        case '?':
            return OPT_ERROR_INVALID;

        case ':':
            return OPT_ERROR_MISSING;

        default:
            action = CONFIG_LIST;
            return OPT_FINAL;
    }
}

int jdfs_config(int argc, char* argv[], const struct conf_data* configuration) {

    snprintf(caller_str, CALLER_STR_BUFSIZE, "jdfs-%s", argv[0]);

    int optchar;
    int old_optind = 1;

    while((optchar = getopt_long(argc, argv, ":hg:ls:u:", longopts, NULL)) != -1) {
        enum optswitch_retval optswitch = config_optswitch(optchar, optarg);

        if (optswitch == OPT_FINAL) {
            break;
        } else if (optswitch == OPT_ERROR_INVALID && optind != old_optind) {
            snprintf(error_str, ERROR_STR_BUFSIZE, "invalid option \"%s\"", argv[optind - 1]);
            return INPUT_ERROR;
        } else if (optswitch == OPT_ERROR_MISSING && optind != old_optind) {
            snprintf(error_str, ERROR_STR_BUFSIZE, "missing argument to \"%s\"", argv[optind - 1]);
            return INPUT_ERROR;
        } else if (optswitch == OPT_ERROR_INVALID && optind == old_optind) {
            snprintf(error_str, ERROR_STR_BUFSIZE, "invalid option \"%s\"", argv[optind]);
            return INPUT_ERROR;
        } else if (optswitch == OPT_ERROR_MISSING && optind == old_optind) {
            snprintf(error_str, ERROR_STR_BUFSIZE, "missing argument to \"%s\"", argv[optind]);
            return INPUT_ERROR;
        }

        // Store the value of optind to check whether it has moved in the next
        // iteration of the loop.
        old_optind = optind;
    }

    switch (action) {
        case CONFIG_HELP:
            config_print_help(argv[0]);
            return SUCCESS;

        case CONFIG_GET:
            return config_get(optarg, configuration->config_path);

        case CONFIG_LIST:
            return print_conf_data(configuration->config_path);

        case CONFIG_SET:
            return config_set(optarg, argv[optind], configuration->config_path);

        case CONFIG_UNSET:
            return config_unset(optarg, configuration->config_path);

        case CONFIG_NULL:
            return ERROR;

        default:
            return print_conf_data(configuration->config_path);
    }
}
